backend/post.js

```diff
@@ -71,45 +71,61 @@ async function getPostJSON() {
 
 	// then try reading file
 	const file = await readFile(POST_EXPORT_FILE, { encoding: "utf8" }).catch(() => null);
-	if (file) return file;
+	if (file) return JSON.parse(file);
 
 	// then generate a new copy if it doesn't exist
 	const json = await generatePostJSON();
 	return json;
 }
 
-router.get("/posts.json", cors(corsOptions), async (_, res) => {
-	const posts = await getPostJSON();
-	if (!posts) return res.status(400).send({ message: "Could not load posts!" }).end();
-	return res.send(JSON.stringify(posts)).end();
-});
-
-// match all routes and continue if no matching post is found
-router.get("*", async (req, res, next) => {
-	const posts = await getPostJSON();
-	// posts couldn't be fetched, go to 404 page
-	if (!posts) next();
-
-	// remove last character (permalinks don't end with a slash)
-	if (req.url.endsWith("/")) req.url = req.url.slice(0, -1);
-	const post = posts[req.url];
-	// post doesn't exist under path, probably another page
-	if (!post) return next();
-
+async function loadPostPage(post) {
 	let data = await readFile(POST_PAGE, { encoding: "utf8" });
 	POST_REPLACE_FIELDS.forEach((token) => {
 		// need to replace br elements with newlines to prevent embeds looking weird
 		data = data.replaceAll(replaceTemplateToken(token), post[token].replaceAll("<br>", "\n"));
 	});
-	// send stringified contents
-	data = data.replaceAll("%postData%", cleanPostData(post));
+	data = data
+		.replaceAll(
+			"%header-img%",
+			post.header_img ||
+				"https://database.faithfulpack.net/images/website/posts/placeholder.jpg",
+		)
+		.replaceAll("%postData%", cleanPostData(post));
+	return data;
+}
+
+// force new generation on startup (prevents new routes not being added)
+const posts = await generatePostJSON();
+
+// match all post routes
+router.get(Object.keys(posts), async (req, res) => {
```

Comment:
```suggestion
// include the ending slash URLs here
const post_urls = Object.keys(posts).map(url => [url, url + '/']).flat();
// to match them here
router.get(post_urls, async (req, res) => {
        // then delete the slash back
```

Reply:
isn't needed, the last slash gets matched by express either way (it just shows up in the query params and the post keys don't include it so we need to remove it)